File: guests/rust/select.rs
//! The select conformance scene: the control owns its selection and
//! reports each change as a 0-based index — the slider contract with
//! an index. The contract is tools/scenes/select.steps.

const OPTIONS: [&str; 3] = ["Red", "Green", "Blue"];

#[derive(Clone)]
enum Msg {
    Picked(usize),
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let picked = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("select");
        let picked = tx.signal("picked: Red");
        let root = tx
            .column(|tx| {
                let color = tx.select(&OPTIONS, 0).id();
                msgs.on_select(color, Msg::Picked);
                tx.label(picked);
            })
            .id();
        tx.mount(root);
        picked
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Picked(index) => {
                ctx.apply(|tx| {
                    tx.write(picked, format!("picked: {}", OPTIONS[index]));
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}

File: guests/haskell/select.hs
{-# LANGUAGE OverloadedStrings #-}

-- The select scene, Haskell port — guests/rust/select.rs,
-- tools/scenes/select.steps.

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

options :: [Text]
options = ["Red", "Green", "Blue"]

main :: IO ()
main = kayaMain $ \app -> do
  _ <- buildTx app $ do
    window primary [WTitle "select"]
    picked <- signal (T.pack "picked: Red")

    let onPick index =
          submitTx app $
            writeSignal picked ("picked: " <> options !! index)

    root <-
      column
        []
        [ selectOn options 0 onPick,
          labelBound picked -- label#0
        ]
    mount root
    return ()
  return ()
